android/app/src/main/java/safetyfirst/androidapp/safetyfirstcontroller/MainActivity.java

```diff
@@ -373,6 +303,196 @@ public void onJoystickMoved(float xPercent, float yPercent, int id) {
+        dialog.show();
+
+
+        newContactPopupSave.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+
+                EmergencyContact contactModel;
+                String regexPattern = "^(.+)@(\\S+)$";
+
+                //Throws exception if added contact information does not meet requirements.
+
+                try {
+                    if (EmergencyContact.patternMatches(newContactPopupEmail.getText().toString(), regexPattern)) {
+                        contactModel = new EmergencyContact(-1, newContactPopupFirstname.getText().toString(), newContactPopupLastname.getText().toString(), Integer.parseInt(newContactPopupMobile.getText().toString()), newContactPopupEmail.getText().toString());
+                        Toast.makeText(MainActivity.this, contactModel.toString(), Toast.LENGTH_SHORT).show();
+                        DataBaseHelper dataBaseHelper = new DataBaseHelper(MainActivity.this);
+                        boolean success = dataBaseHelper.addOne(contactModel);
+                        Toast.makeText(MainActivity.this, "Contact Added", Toast.LENGTH_SHORT).show();
+                    } else {
+                        throw new Exception();
+                    }
+
+                } catch (Exception e) {
+                    Toast.makeText(MainActivity.this, "Error creating contact", Toast.LENGTH_SHORT).show();
+                    contactModel = new EmergencyContact(-1, "error", "error", 0, "error");
+                }
+
+                dialog.dismiss();
+
+            }
+        });
+
+        newContactPopupCancel.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                //define cancel button
+                dialog.dismiss();
+            }
+        });
+    }
+
+    public void openContactActivity() {
+        Intent intent = new Intent(this, ContactList.class);
+        startActivity(intent);
+        lv_contactList = findViewById(R.id.lv_contactList);
+    }
+
+
+    public void sendMessageEmergencyContact() {
+
+        DataBaseHelper phone_number_data = new DataBaseHelper(this);
+
+        try {
+            String query = "SELECT CONTACT_PHONE_NUMBER FROM CONTACT_TABLE ORDER BY CONTACT_PHONE_NUMBER DESC LIMIT 1";
+            SQLiteDatabase dbs = phone_number_data.getReadableDatabase();
+            Cursor result = dbs.rawQuery(query, null);
+            result.moveToFirst();
+            int phone_number = result.getInt(result.getColumnIndexOrThrow("CONTACT_PHONE_NUMBER"));
+            Intent intent = new Intent(Intent.ACTION_SENDTO);
+            intent.setData(Uri.parse("sms:" + phone_number));
+            startActivity(intent);
+
+        } catch (Exception e) {
+            Toast.makeText(MainActivity.this, "No emergency contact added", Toast.LENGTH_SHORT).show();
+        }
+
+    }
+
+
+    public void callEmergencyContact() {
```

Comment:
Looks very similar to method right above, has the same naming conventions issues,
also maybe it could be made into one method?